datafusion/spark/src/function/misc/bitmap_count.rs

```diff
@@ -0,0 +1,175 @@
+// specific language governing permissions and limitations
+// under the License.
+
+use std::any::Any;
+use std::sync::Arc;
+
+use arrow::array::{
+    Array, ArrayRef, BinaryArray, BinaryViewArray, FixedSizeBinaryArray, Int64Array,
+    LargeBinaryArray,
+};
+use arrow::datatypes::DataType;
+use arrow::datatypes::DataType::{
+    Binary, BinaryView, FixedSizeBinary, Int64, LargeBinary,
+};
+use datafusion_common::utils::take_function_args;
+use datafusion_common::Result;
+use datafusion_common::{exec_err, internal_datafusion_err, DataFusionError};
+use datafusion_expr::function::Hint;
+use datafusion_expr::ScalarFunctionArgs;
+use datafusion_expr::{ColumnarValue, ScalarUDFImpl, Signature, Volatility};
+use datafusion_functions::utils::make_scalar_function;
+use datafusion_functions::{downcast_arg, downcast_named_arg};
+
+#[derive(Debug)]
+pub struct BitmapCount {
+    signature: Signature,
+}
+
+impl Default for BitmapCount {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl BitmapCount {
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::any(1, Volatility::Immutable),
+        }
+    }
+}
+
+impl ScalarUDFImpl for BitmapCount {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "bitmap_count"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
+        match arg_types.first() {
+            Some(Binary | BinaryView | FixedSizeBinary(_) | LargeBinary) => Ok(Int64),
+            _ => Err(DataFusionError::Internal(format!(
+                "{} should only be called with a binary",
+                self.name(),
+            ))),
+        }
+    }
+
+    fn invoke_with_args(&self, args: ScalarFunctionArgs) -> Result<ColumnarValue> {
+        make_scalar_function(bitmap_count_inner, vec![Hint::AcceptsSingular])(&args.args)
+    }
+}
+
+fn binary_count_ones(opt: Option<&[u8]>) -> Option<i64> {
+    opt.map(|value| value.iter().map(|b| b.count_ones() as i64).sum())
+}
+
+macro_rules! downcast_and_count_ones {
+    ($input_array:expr, $array_type:ident) => {{
+        let arr = downcast_arg!($input_array, $array_type);
+        Ok(arr.iter().map(binary_count_ones).collect::<Int64Array>())
+    }};
+}
+
+pub fn bitmap_count_inner(arg: &[ArrayRef]) -> Result<ArrayRef> {
+    let [input_array] = take_function_args("bitmap_count", arg)?;
+
+    let res: Result<Int64Array> = match &input_array.data_type() {
+        Binary => downcast_and_count_ones!(input_array, BinaryArray),
+        BinaryView => downcast_and_count_ones!(input_array, BinaryViewArray),
+        LargeBinary => downcast_and_count_ones!(input_array, LargeBinaryArray),
+        FixedSizeBinary(_size) => {
+            downcast_and_count_ones!(input_array, FixedSizeBinaryArray)
+        }
+        array_type => exec_err!("`bitmap_count` does not support type '{array_type:?}'."),
```

Comment:
I notice in `bit_count` we return a plan err instead:

https://github.com/apache/datafusion/blob/f87763dc9ffb5ce97482002ae796beba514666ad/datafusion/spark/src/function/bitwise/bit_count.rs#L141-L146

Might need to decide which type of error to return and if we can make the messages consistent 🤔

Reply:
placed `exec_err` here because it is in the execution level
but this would be really unreachable at the execution level because `return_type` will already fail at the plan level, so I think that `plan_err` would fit better, will replace this to `plan_err`

Reply:
Sounds good 👍